### ISSUE TITLE
[feature] aliases should be run immediately and get a file object as stdin

### ISSUE THREAD

**Issue by azag0** [xonsh/xonsh#831]
Currently, the second argument of aliases is a string and the function is called only when the previous process or alias finishes. So one cannot reimplement `cat` in xonsh properly, for example. Could the aliases be dispatched immediately in a separate thread or process (as real subprocesses seem to be?) and obtain stdin as a file object? That way, one could yield from/iterate over the file and get the desired async behaviour.

Implementing it using processes would be probably slower than threads but might be simpler regarding signal handling.

Also, could we have something along these lines as a builtin?

``` python
def alias(f):
    assert f.__name__[0] == '_'
    aliases[f.__name__[1:]] = f
    return f
```

At least to me it seems that in xonsh, I'll be defining aliases all the time to pipe output of python functions to subprocesses. Or is there a better mechanism in place for that?

**Comment by azag0** [comment]
Actually, what really happens when I run something like this in xonsh?
```
🐌 11:29 her@air ~ bash -c 'echo hey; sleep 2' | mycat | bash -c 'echo test; cat; sleep 2'
```
I see that both bash processes run right ahead, and that the command produces the correct output, but what get's connected to the second bash's stdin when it's launched?

**Comment by azag0** [comment]
Ah, ok, I see that this capability is all here, but it's the doing of `SimpleProcProxy`. So how can one make an alias that get's treated as the full `ProcProxy`?

**Comment by adqm** [comment]
As of right now, it should just involve making a function with a different signature (`run_subproc` looks at how many arguments the function takes to determine which type to use).  So instead of making a function `my_alias(args, stdin_string)`, you would make `my_alias(args, stdin, stdout, stderr)`.

In that second form (which the first one gets converted to anyway), the last three arguments are all file-like objects.

**Comment by azag0** [comment]
I see. Nice. Xonsh nice. But it still doesn't seem to work like expected. Can you reproduce this? Note the pauses.
```
🐌 15:14 her@air ~ @alias
................. def _mycat(args, stdin, stdout, stderr):
.................     for l in stdin:
.................         stdout.write(l)
.................         
🐌 15:14 her@air ~ bash -c 'echo hey; sleep 2' | cat | bash -c 'cat; sleep 2'
[no wait]
hey
[wait 4 s]
🐌 15:14 her@air ~ bash -c 'echo hey; sleep 2' | mycat | bash -c 'cat; sleep 2'
[wait 2 s]
hey
[wait 2 s]
🐌 15:14 her@air ~ bash -c 'echo hey; sleep 2' | mycat
[no wait]
hey
[no wait]
Exception in thread Thread-371:
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.5/site-packages/xonsh/proc.py", line 173, in run
    r = self.f(self.args, sp_stdin, sp_stdout, sp_stderr)
  File "<xonsh-code>", line 3, in _mycat
ValueError: I/O operation on closed file.
```
And this gets more interesting.
```
🐌 15:17 her@air ~ @alias
................. def _mycat(args, stdin, stdout, stderr):
.................     for l in stdin:
.................         sys.stdout.write(repr(l) + '\n')
.................         stdout.write(repr(l) + '\n')
.................         
🐌 15:18 her@air ~ bash -c 'echo hey; sleep 2' | mycat | bash -c 'cat; sleep 2'
'hey\n'
Exception in thread Thread-427:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/xonsh/built_ins.py", line 512, in run_subproc
    **subproc_kwargs)
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/subprocess.py", line 950, in __init__
    restore_signals, start_new_session)
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/subprocess.py", line 1544, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: '+'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.5/site-packages/xonsh/proc.py", line 173, in run
    r = self.f(self.args, sp_stdin, sp_stdout, sp_stderr)
  File "<xonsh-code>", line 5, in _mycat
  File "/usr/local/lib/python3.5/site-packages/xonsh/built_ins.py", line 615, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/usr/local/lib/python3.5/site-packages/xonsh/built_ins.py", line 520, in run_subproc
    if len(sug.strip()) > 0:
AttributeError: 'NoneType' object has no attribute 'strip'
```
Seems like xonsh is trying to parse what gets into stdout of mycat as a xonsh code? Or something like that.

**Comment by scopatz** [comment]
@azag0 a decorator that registers aliases seems pretty useful to me.  Please feel free to PR it in.  I am not 100% convinced that it should be a builtin on its own.  Maybe it could be a static method on the Aliases class?  Then it would be available everywhere as `@aliases.reg` or whatever you want to name it.

**Comment by scopatz** [comment]
Ok, I can reproduce the I/O error

**Comment by scopatz** [comment]
I can reproduce them all, actually.

**Comment by scopatz** [comment]
So you can reproduce the I/O error with simply:

```python
def _nullloop(args, stdin, stdout, stderr):
    for l in stdin:
        pass
```

Also, the following just quits immediately:

```python
aliases['null'] = lambda args, stdin, stdout, stderr: None
```

**Comment by scopatz** [comment]
This may be relevant.

**Comment by adqm** [comment]
The IO error is a timing thing, and one that I think I handle on #742.  The behavior on that branch seems to be closer to matching bash, though still not perfect.

I can separate out the relevant parts of that PR into a second one if that's helpful (or, now that work is winding down for the semester, maybe I can just wrap up that whole PR).

**Comment by azag0** [comment]
btw, when this is in, xonsh will have another take on fish. fish cannot job control/fork its own code either (happy fish user here)

**Comment by gforsyth** [comment]
@azag0 -- I can no longer reproduce this error on master.  Could you try it out and let us know if it's still an issue?  Thanks!

**Comment by azag0** [comment]
Yes, I'm still getting the same two errors with fresh `pip3 install git+https://github.com/scopatz/xonsh.git`.

**Comment by scopatz** [comment]
Yeah, I think that this one might need some work still, since it isn't about job control per-se

**Comment by scopatz** [comment]
I believe that this issue has been addressed by the subprocess refactor in v0.5+. Please feel free to reopen or open an new issue if you continue to experience problems.
